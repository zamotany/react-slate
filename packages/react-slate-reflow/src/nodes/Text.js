/* @flow */

import assert from 'assert';
import UnitLayout from '../layout/elements/UnitLayout';
import type { Node } from '../types';

export default class Text implements Node {
  body: string = '';
  parent = null;
  // Children should always be empty.
  children = Object.freeze([]);
  layout = new UnitLayout(this);

  setBody(body: string) {
    assert(!body.includes('\n'), 'Text body cannot have new line characters');
    this.body = body;
  }
}
