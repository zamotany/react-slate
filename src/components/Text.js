// @flow

import React, { Children } from 'react';
import chalk from 'chalk';
import { Chunk, Endl } from './';

type Style = {|
  color?: string,
  backgroundColor?: string,
  fontWeight?: 'bold' | 'normal',
  fontStyle?: 'italic' | 'normal',
  textDecoration?: 'underline' | 'line-through' | 'normal',
  textTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase',
  visibility?: 'visible' | 'hidden',
|};

type Props = {
  style?: Style,
  endl?: boolean,
  children: any,
};

function capitalize(text: string) {
  return text
    .split(' ')
    .map(t => t.charAt(0).toUpperCase() + t.slice(1))
    .join(' ');
}

function stylize(style: Style, text: string) {
  /* eslint-disable no-param-reassign */

  let color = style.color;
  let inverse = false;

  if (style.backgroundColor) {
    color = style.backgroundColor;
    inverse = true;
  }

  let enhancer = chalk.reset;

  if (color) {
    if (color.startsWith('#')) {
      enhancer = enhancer.hex(color);
    } else if (color.startsWith('rgb')) {
      const rgb = color
        .replace(/[rgb()]/g, '')
        .split(',')
        .map(i => parseInt(i, 10));
      enhancer = chalk.rgb(...rgb);
    } else {
      enhancer = chalk.keyword(color);
    }
  }

  if (inverse) {
    enhancer = enhancer.inverse;
  }

  if (style.fontWeight === 'bold') {
    enhancer = enhancer.bold;
  }

  if (style.fontStyle === 'italic') {
    enhancer = enhancer.italic;
  }

  if (style.textDecoration) {
    // eslint-disable-next-line default-case
    switch (style.textDecoration) {
      case 'underline':
        enhancer = enhancer.underline;
        break;
      case 'line-through':
        enhancer = enhancer.strikethrough;
        break;
    }
  }

  if (style.visibility === 'hidden') {
    enhancer = enhancer.hidden;
  }

  if (style.textTransform) {
    // eslint-disable-next-line default-case
    switch (style.textTransform) {
      case 'uppercase':
        text = text.toUpperCase();
        break;
      case 'lowercase':
        text = text.toLowerCase();
        break;
      case 'capitalize':
        text = capitalize(text);
        break;
    }
  }

  return enhancer(text);
}

export default function Text(props: Props) {
  const { children, style, endl } = props;

  return (
    <Chunk>
      {style
        ? Children.map(children, child => {
            if (typeof child === 'string' || typeof child === 'number') {
              return stylize(style, String(child));
            }

            if (child && child.type === Text) {
              return React.cloneElement(child, {
                style: child.props.style
                  ? { ...style, ...child.props.style }
                  : style,
              });
            }

            return child;
          })
        : children}
      {endl ? <Endl /> : null}
    </Chunk>
  );
}
