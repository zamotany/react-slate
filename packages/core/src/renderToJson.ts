import Reconciler from 'react-reconciler';
import { inspect } from 'util';
import createReconcilerConfig from './host/reconcilerConfig';
import View from './host/nodes/View';
import Text from './host/nodes/Text';
import { Layout } from './layout';
import { JsonView, JsonText, JsonParagraph } from './types';
import { AsyncIterator } from './utils';
import Paragraph from './host/nodes/Paragraph';

function traverse(
  node: View | Paragraph | Text,
  layout: Layout
): JsonView | JsonParagraph | JsonText | undefined {
  if (node instanceof View) {
    return {
      node: 'View',
      children: node.children.map((child, index) =>
        traverse(child, layout.child(index))
      ),
      width: layout.width,
      height: layout.height,
      x: layout.x,
      y: layout.y,
      style: node.style,
    } as JsonView;
  } else if (node instanceof Text) {
    return {
      node: 'Text',
      body: node.getBody(),
      width: layout.width,
      height: layout.height,
      x: layout.x,
      y: layout.y,
      style: node.style,
    } as JsonText;
  } else if (node instanceof Paragraph) {
    return {
      node: 'Paragraph',
      children: node.children.map((child, index) =>
        traverse(child, layout.child(index))
      ),
      width: layout.width,
      height: layout.height,
      x: layout.x,
      y: layout.y,
      style: node.style,
    } as JsonParagraph;
  }

  throw new Error(`Unknown instance of node: ${inspect(node)}`);
}

export default function renderToJson(
  element: JSX.Element,
  {
    width,
    height,
    maxRenders,
  }: { width?: number; height?: number; maxRenders?: number } = {}
): {
  snapshot: JsonView | undefined;
  start: () => Iterable<Promise<JsonView | undefined>>;
  stop: () => void;
} {
  let currentSnapshot: JsonView | undefined;
  const asyncIterator = new AsyncIterator<JsonView | undefined>(maxRenders);
  const container = new View();

  container.setLayoutStyle({ width: '100%', height: '100%' });
  const reconciler = Reconciler(
    createReconcilerConfig(container, () => {
      const layout = container.layoutNode.computeLayout({
        width: width || null,
        height: height || null,
      });
      container.notifyOnLayoutHook(layout, { offsetX: 0, offsetY: 0 });
      currentSnapshot = traverse(container, layout) as JsonView;
      asyncIterator.nextValue(currentSnapshot);
    })
  );
  const node = reconciler.createContainer(container, false, false);
  reconciler.updateContainer(element, node, null, () => undefined);

  return {
    get snapshot() {
      return currentSnapshot;
    },
    start: () => asyncIterator.makeIterator({}),
    stop: () => {
      asyncIterator.finish(undefined);
      reconciler.updateContainer(null, node, null, () => undefined);
    },
  };
}
