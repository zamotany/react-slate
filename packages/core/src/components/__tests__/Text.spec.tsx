import React from 'react';
import Text from '../Text';
import renderToJson from '../../renderToJson';

describe('Text component', () => {
  it('should transform children', () => {
    expect(
      renderToJson(<Text textTransform="uppercase">hello</Text>).snapshot
    ).toMatchSnapshot();
    expect(
      renderToJson(<Text textTransform="lowercase">HELLO</Text>).snapshot
    ).toMatchSnapshot();
    expect(
      renderToJson(<Text textTransform="capitalize">hello world</Text>).snapshot
    ).toMatchSnapshot();
    expect(
      renderToJson(<Text textTransform="none">hello</Text>).snapshot
    ).toMatchSnapshot();
    expect(
      renderToJson(<Text textTransform="uppercase">{['hello ', 'world']}</Text>)
        .snapshot
    ).toMatchSnapshot();
  });

  it('should add modifiers', () => {
    expect(
      renderToJson(<Text>hello</Text>).snapshot.children[0].style.modifiers
    ).toBeUndefined();
    expect(
      renderToJson(<Text bold>hello</Text>).snapshot.children[0].style.modifiers
    ).toEqual(['bold']);
    expect(
      renderToJson(<Text dim>hello</Text>).snapshot.children[0].style.modifiers
    ).toEqual(['dim']);
    expect(
      renderToJson(<Text underline>hello</Text>).snapshot.children[0].style
        .modifiers
    ).toEqual(['underline']);
    expect(
      renderToJson(<Text lineThrough>hello</Text>).snapshot.children[0].style
        .modifiers
    ).toEqual(['strikethrough']);
    expect(
      renderToJson(<Text italic>hello</Text>).snapshot.children[0].style
        .modifiers
    ).toEqual(['italic']);

    expect(
      renderToJson(
        <Text bold italic>
          hello
        </Text>
      ).snapshot.children[0].style.modifiers
    ).toEqual(['bold', 'italic']);
  });

  it('should set color and bgColor', () => {
    expect(
      renderToJson(
        <Text color="red" bgColor="blue">
          hello
        </Text>
      ).snapshot.children[0].style
    ).toEqual({ color: 'red', bgColor: 'blue', modifiers: undefined });
  });
});
