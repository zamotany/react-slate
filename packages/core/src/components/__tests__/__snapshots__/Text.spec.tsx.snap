// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component should transform children 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "HELLO",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "width": 5,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 5,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "hello",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "width": 5,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 5,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 3`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "Hello World",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 11,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "width": 11,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 11,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 4`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "hello",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "width": 5,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 5,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 5`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "HELLO ",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 6,
          "x": 0,
          "y": 0,
        },
        Object {
          "body": "WORLD",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 6,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "width": 11,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 11,
  "x": 0,
  "y": 0,
}
`;
